doc/spec/package-lock.md

```diff
@@ -0,0 +1,290 @@
+# package-lock and npm-shrinkwrap
+
+`npm` can have one of two different lock files:
+
+* `package-lock.json`, which is ordinarily always present and is never published.
+* `npm-shrinkwrap.json`, which is created with `npm shrinkwrap` and usually published.
+
+You can only have one of them and in the event that you have both,
+`npm-shrinkwrap.json` takes precedence.  The files are exactly the same
+format and in fact all the `npm shrinkwrap` command does is rename your
+`package-lock.json`.
+
+Through the rest of this document we will refer to the package-lock and
+`package-lock.json` but everything also applies to `npm-shrinkwrap.json`.
+
+## File Format
+
+### name
+
+The name of the package this is a package-lock for. This must match what's in `package.json`.
+
+### version
+
+The version of the package this is a package-lock for. This must match what's in `package.json`.
+
+### createdWith *(new)*
+
+The `name@version` of the creator of this package-lock.  The name part must be
+a valid npm package name.  The version part must be a valid semver version
+string. For example: `npm@5.0.0`
+
+This is non-normative and intended to help in isolating the cause of errors.
+
+### lockfileVersion *(new)*
+
+An integer version, starting at `1` with the version number of this document
+whose semantics were used when generating this `package-lock.json`.
+
+### packageIntegrity *(new)*
```

Comment:
why not just duplicating the specifiers in `package-lock`? Something like

```
specs: {
  foo: '^1.0.0',
  bar: '2.0.0'
}
```

dependencies, devDependencies, optionalDependencies can be merged into one property. In that case, package managers will know exactly which dependencies should be updated and skip those which did not change. It can also print some info message that "this or that does not match the spec from `package.json`"